core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -237,13 +237,17 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
         dynamicConfigManager.startup()
 
         /* tell everyone we are alive */
-        val listeners = config.advertisedListeners.map {case(protocol, endpoint) =>
-          if (endpoint.port == 0)
-            (protocol, EndPoint(endpoint.host, socketServer.boundPort(protocol), endpoint.protocolType))
-          else
-            (protocol, endpoint)
+        val advertisedListeners = config.advertisedListeners.mapValues { endpoint =>
+          val host =
+            if (Option(endpoint.host).map(_.trim).forall(host => host.isEmpty || host == "0.0.0.0"))
+              InetAddress.getLocalHost.getCanonicalHostName
```

Comment:
@harshach Thanks for the comment.
I understand your point. However, I guess using `InetAddress.getLocalHost()` itself shouldn't be a problem since we're already doing it if host omitted in listeners: https://github.com/apache/kafka/pull/1816/files/0eec0393b41e4e75d942b3431e328d5acc18ca7f#diff-8634f708a0a9a5eea3c1b79c30a82c26L60

Even though we fallback to default interface when empty host or "0.0.0.0" is given to `advertised.listeners`, users can still control an address to be published on ZK through setting `advertised.listeners` explicitly.

Then when users supply configuration like below, I think fallback to default interface for advertised address is the expected behavior for most users IMHO:
```
listeners=PLAINTEXT://0.0.0.0:9092
# advertised.listeners=
```

WDYT?